dialogue-blocking-channels/src/main/java/com/palantir/dialogue/blocking/BlockingChannelAdapter.java

```diff
@@ -52,42 +57,66 @@ private BlockingChannelAdapter() {}
     private static final class BlockingChannelAdapterChannel implements Channel {
 
         private final BlockingChannel delegate;
-        private final ListeningExecutorService executor;
+        private final ExecutorService executor;
 
-        BlockingChannelAdapterChannel(BlockingChannel delegate, ListeningExecutorService executor) {
+        BlockingChannelAdapterChannel(BlockingChannel delegate, ExecutorService executor) {
             this.delegate = delegate;
             this.executor = executor;
         }
 
         @Override
         public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
-            return executor.submit(new BlockingChannelAdapterTask(delegate, endpoint, request));
+            SettableFuture<Response> result = SettableFuture.create();
+            BlockingChannelAdapterTask runnable = new BlockingChannelAdapterTask(delegate, endpoint, request, result);
+            Future<?> future = executor.submit(runnable);
+            // The executor task should be interrupted on termination
+            result.addListener(() -> future.cancel(true), MoreExecutors.directExecutor());
+            return result;
         }
 
         @Override
         public String toString() {
             return "BlockingChannelAdapterChannel{delegate=" + delegate + ", executor=" + executor + '}';
         }
 
-        /**
-         * Callable to run a blocking request. This could be a lambda,
-         * but concrete classes provide better stack traces.
-         */
-        static final class BlockingChannelAdapterTask implements Callable<Response> {
+        static final class BlockingChannelAdapterTask implements Runnable {
 
+            private final SettableFuture<Response> result;
             private final BlockingChannel delegate;
             private final Endpoint endpoint;
             private final Request request;
 
-            BlockingChannelAdapterTask(BlockingChannel delegate, Endpoint endpoint, Request request) {
+            BlockingChannelAdapterTask(
+                    BlockingChannel delegate, Endpoint endpoint, Request request, SettableFuture<Response> result) {
+                this.result = result;
                 this.delegate = delegate;
                 this.endpoint = endpoint;
                 this.request = request;
             }
 
             @Override
-            public Response call() throws IOException {
-                return delegate.execute(endpoint, request);
+            public void run() {
+                if (result.isDone()) {
+                    return;
```

Comment:
Tiny bit confusing that we're using isDone as a proxy for `isCancelled`... I think a little comment would be helpful here, something along the lines of `if a user cancels the SettableFuture we exposed, then there's no need to call the delegate`.

Reply:
actually how could this ever be set to a value already (but not cancelled) here? isn't that an invariant violation?

Reply:
`isCancelled` works just as well, I chose isDone because it maps more closely to the final check we make on completion, settableFuture.set returns false if the future is done (which includes isCancelled).